### PR TITLE
Fixup central config

### PR DESCRIPTION
**Pull request by redp01**
@vuldin - moving the commits that were on your branch for PR85 onto v2.

Essentially this adds:

- upgrade to 22.1.4 and with it the need to use central config. 
- post-upgrade hook is also added to ensure that changes to the cluster values can be updated correctly by rpk with a helm update.
- sasl configuration has changed between versions, so amended this.
- The amendment to sasl configuration also highlighted that the helm tests were broken since a previous PR adjusted the TLS configuration. These have been fixed.
- Ordered Ready deployment used to facilitate lower down time upgrades

@vuldin - would be great if you could please review and assist in test cases for upgrades.
Questions: 
the values files between versions differ - i.e. sub headings for cluster, node, tuneables. Do we favour one format or handle both?
Updating the config triggers redeployment though one of the advantages of central config updates is often no restart necessary? Do we want to handle this in a helm update?

### CODE REVIEW

File: redpanda/templates/configmap.yaml [hunk 1]
```diff
@@ -30,6 +30,21 @@ metadata:
...
+      superusers: ['{{ (first .).name }}']
+  {{- end }}
+{{- end }}
+{{- end }}
+ {{- with .Values.config.cluster }}
+    {{- toYaml . | nindent 6 }}
+  {{- end }}
+  {{- with .Values.config.tunable }}
+    {{- toYaml . | nindent 6 }}
+  {{- end }}
```
Comment (vuldin) [inline redpanda/templates/configmap.yaml:47]: It seems that `bootstrap.yaml` will be configured here with values from `values.yaml` regardless of the Redpanda version. I would expect to be able to use Redpanda `21.x` by modifying the version, and then the helm chart add the config to `redpanda.yaml` instead `bootstrap.yaml`.

File: redpanda/templates/post-upgrade.yaml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+{{- else if eq (include "tls-enabled" .) "true" }}
+            rpk --api-urls {{ template "redpanda.name" . }}-0.{{ include "redpanda.internal.domain" . }}:{{ template "redpanda.admin.port" $ }} cluster config import -f /tmp/base-config/bootstrap.yaml --admin-api-tls-enabled --admin-api-tls-truststore /etc/tls/certs/admin/ca.crt --tls-enabled --tls-truststore /etc/tls/certs/admin/ca.crt;
+{{- else if eq (include "tls-enabled" .) "true" }}
+            rpk --api-urls {{ template "redpanda.name" . }}-0.{{ include "redpanda.internal.domain" . }}:{{ template "redpanda.admin.port" $ }} cluster config import -f /tmp/base-config/bootstrap.yaml --user admin --password test --sasl-mechanism SCRAM-SHA-256;
+{{- else }}
+            rpk --api-urls {{ template "redpanda.name" . }}-0.{{ include "redpanda.internal.domain" . }}:{{ template "redpanda.admin.port" $ }} cluster config import -f /tmp/base-config/bootstrap.yaml;
+{{- end }}
```
Comment (vuldin) [inline redpanda/templates/post-upgrade.yaml:50]: If I understand, this reconfigures the admin URLs to make use of TLS and/or SASL after the cluster has been started. This allows these auth features to be enabled and then applied after a cluster has already been created.

If that's the case, then what about the other listeners which may also make use of TLS/SASL? Would they also require modification at this point (and if not they would be in a broken state)?

File: redpanda/values.yaml [hunk 3]
```diff
@@ -316,94 +316,104 @@ serviceAccount:
 # These values do not pertain to the kubernetes objects created with helm.
 # Instead these parameters get passed directly to the Redpanda binary at startup.
 # See https://docs.redpanda.com/docs/cluster-administration/configuration/
+# There are three subsections, cluster, tunable and node. The cluster and tunable sections
+# are written to the cluster .bootstrap.yaml and read into central config upon startup.
```
Comment (vuldin) [inline redpanda/values.yaml:320]: Maybe this text should be taken out since it doesn't make sense for all Redpanda versions (older version wouldn't have `bootstrap.yaml` and would expect config to be written to `redpanda.yaml`).

**Review by vuldin — CHANGES_REQUESTED** [review]
I mainly had a question around how to handle `values.yaml` config with previous versions. The PR is looking good, thanks!

File: redpanda/templates/configmap.yaml (hunk 1 above)
Comment (redp01) [reply redpanda/templates/configmap.yaml:47]: @vuldin - I literally just added that question to the PR comments. It should be possible to alter the config map so depending on version either updates the configmap for redpanda.yaml or not. I will look into it now.

File: redpanda/templates/post-upgrade.yaml (hunk 2 above)
Comment (redp01) [reply redpanda/templates/post-upgrade.yaml:50]: @vuldin - so what this does is apply the cluster config and tuneables transitively from the changes to the values.yaml. It switches based on the tls / sasl mode in use. So its not changing the auth features - its just trying to apply the new config via the admin config import method which may be configured for tls, sasl or not.

**Comment by vuldin** [comment]
It looks like this PR may fix #109 , fix #110 (let me know if you think this isn't the case).

**Comment by redp01** [comment]
> It looks like this PR may fix #109 and #110 (let me know if you think this isn't the case).

Yes. #109 and generally #110 though we need to test with the individual service overrides - for example we could enable overarching TLS then set enabled: false for one service. Though this is certainly better as I would imagine its unlikely to enable TLS for one service and not all.

**Comment by rockdata01** [comment]
> Also closes #54

Yes

File: redpanda/templates/configmap.yaml (hunk 1 above)
Comment (rockdata01) [reply redpanda/templates/configmap.yaml:46]: Commit [b6e8ff8](https://github.com/redpanda-data/helm-charts/pull/117/commits/b6e8ff8e5ded707999865a629155e5ae8394e2c3) should now allow the old style values.yaml to be compatible with the older versions.

File: redpanda/values.yaml (hunk 3 above)
Comment (rockdata01) [reply redpanda/values.yaml]: @vuldin - the additional and potentially confusing lines are now removed in [24e9ee1](https://github.com/redpanda-data/helm-charts/pull/117/commits/24e9ee1aa87bcf44d9c39e677403c3e3c626ab21)

File: redpanda/templates/configmap.yaml [hunk 4]
```diff
@@ -30,6 +30,21 @@ metadata:
...
+      superusers: ['{{ (first .).name }}']
+  {{- end }}
+{{- end }}
+ {{- with .Values.config.cluster }}
+    {{- toYaml . | nindent 6 }}
+  {{- end }}
+  {{- with .Values.config.tunable }}
+    {{- toYaml . | nindent 6 }}
+  {{- end }}
+{{- end }}
```
Comment (vuldin) [inline redpanda/templates/configmap.yaml:47]: I think there should be an `else` here, since we would want to take a different path for lower semver values (Redpanda versions without central config support). In this new `else` branch, `.Values.config.cluster` and `.Values.config.tunable` would be written to `redpanda.yaml` in the appropriate place.

File: redpanda/templates/configmap.yaml [hunk 5]
```diff
@@ -30,6 +30,21 @@ metadata:
   {{- toYaml . | nindent 4 }}
 {{- end }}
 data:
+{{- if (include "redpanda.semver" . | semverCompare ">=22.1.4") }}
```
Comment (vuldin) [inline redpanda/templates/configmap.yaml:33]: I believe the correct version would be `22.1.1`: https://github.com/redpanda-data/redpanda/releases/tag/v22.1.1

File: redpanda/templates/configmap.yaml [hunk 6]
```diff
@@ -39,30 +54,35 @@ data:
...
-{{- if .Values.auth.sasl.enabled }}
-  {{- with .Values.auth.sasl.users }}
+{{- if not (include "redpanda.semver" . | semverCompare ">=22.1.4") }}
+  {{- if .Values.auth.sasl.enabled }}
+    {{- with .Values.auth.sasl.users }}
       superusers:
     {{ toYaml . | indent 8 }}
+    {{- end }}
+  {{- end }}
+  {{- if not (eq (toYaml .Values.config) "null") }}
```
Comment (vuldin) [inline redpanda/templates/configmap.yaml:64]: This isn't a request to change, but is this the same as the following line?
```
{{- if not (hasKey .Values.listeners.rpc "tls") }}
```

File: redpanda/templates/post-upgrade.yaml [hunk 7]
```diff
@@ -0,0 +1,85 @@
+{{- if (include "redpanda.semver" . | semverCompare ">=22.1.4") }}
```
Comment (vuldin) [inline redpanda/templates/post-upgrade.yaml:1]: Same comment about semver version as before, I believe this should be `22.1.1`.

File: redpanda/templates/post-upgrade.yaml [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        image: {{ .Values.image.repository }}:{{ .Values.image.tag }}
+        command: ["/bin/sh", "-c"]
+        args:
+          - >
+{{- if and (eq (include "tls-enabled" .) "true")  (eq (include "sasl-enabled" .) "true") }}
+            rpk --api-urls {{ template "redpanda.name" . }}-0.{{ include "redpanda.internal.domain" . }}:{{ template "redpanda.admin.port" $ }} cluster config import -f /tmp/base-config/bootstrap.yaml --admin-api-tls-enabled --admin-api-tls-truststore /etc/tls/certs/admin/ca.crt --tls-enabled --tls-truststore /etc/tls/certs/admin/ca.crt --user admin --password test --sasl-mechanism SCRAM-SHA-256;
+
+{{- else if eq (include "tls-enabled" .) "true" }}
+            rpk --api-urls {{ template "redpanda.name" . }}-0.{{ include "redpanda.internal.domain" . }}:{{ template "redpanda.admin.port" $ }} cluster config import -f /tmp/base-config/bootstrap.yaml --admin-api-tls-enabled --admin-api-tls-truststore /etc/tls/certs/admin/ca.crt --tls-enabled --tls-truststore /etc/tls/certs/admin/ca.crt;
+{{- else if eq (include "tls-enabled" .) "true" }}
```
Comment (vuldin) [inline redpanda/templates/post-upgrade.yaml:46]: Aren't lines 44 and 46 identical, which would mean line 44 would potentially pass to line 45 and then line 46 would never pass to line 47?

File: redpanda/templates/post-upgrade.yaml [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: {{ template "redpanda.name" . }}-post-upgrade
+        image: {{ .Values.image.repository }}:{{ .Values.image.tag }}
+        command: ["/bin/sh", "-c"]
+        args:
+          - >
+{{- if and (eq (include "tls-enabled" .) "true")  (eq (include "sasl-enabled" .) "true") }}
+            rpk --api-urls {{ template "redpanda.name" . }}-0.{{ include "redpanda.internal.domain" . }}:{{ template "redpanda.admin.port" $ }} cluster config import -f /tmp/base-config/bootstrap.yaml --admin-api-tls-enabled --admin-api-tls-truststore /etc/tls/certs/admin/ca.crt --tls-enabled --tls-truststore /etc/tls/certs/admin/ca.crt --user admin --password test --sasl-mechanism SCRAM-SHA-256;
```
Comment (vuldin) [inline redpanda/templates/post-upgrade.yaml:42]: I see this job run `rpk cluster config import /tmp/base-config/bootstrap.yaml`. I don't remember why this needs to be ran though, do you have details? I will look more into it tomorrow also but wanted to make a comment to remember.

File: redpanda/templates/statefulset.yaml [hunk 10]
```diff
@@ -70,6 +70,9 @@ spec:
               CONFIG=/etc/redpanda/redpanda.yaml;
               NODE_ID=${SERVICE_NAME##*-};
               cp /tmp/base-config/redpanda.yaml "$CONFIG";
+              {{- if (include "redpanda.semver" . | semverCompare ">=22.1.4") }}
```
Comment (vuldin) [inline redpanda/templates/statefulset.yaml:73]: Same as previous comments: semver should be `22.1.1`

**Review by vuldin — CHANGES_REQUESTED** [review]
Thanks for the update to this PR, it's starting to look good. I need to do more testing locally running the code, but I went ahead and added my questions and comments so you would have them at the start of your day. Let me know if you have any questions/comments.

File: redpanda/templates/configmap.yaml (hunk 1 above)
Comment (vuldin) [reply redpanda/templates/configmap.yaml:46]: Discussion continued [here](https://github.com/redpanda-data/helm-charts/pull/117#discussion_r929542045)

File: redpanda/templates/post-upgrade.yaml (hunk 9 above)
Comment (vuldin) [reply redpanda/templates/post-upgrade.yaml:42]: I found the answer to this question: `.bootstrap.yaml` applies config on first start only, while `rpk cluster config import` can be set config for a running cluster (and does so through the admin API).

File: redpanda/templates/statefulset.yaml (hunk 10 above)
Comment (redp01) [reply redpanda/templates/statefulset.yaml]: Fixed in [519e4f1](https://github.com/redpanda-data/helm-charts/pull/117/commits/519e4f193b60c4e2410f2b1042c3eb2d958f5ea7)

File: redpanda/templates/post-upgrade.yaml (hunk 7 above)
Comment (redp01) [reply redpanda/templates/post-upgrade.yaml]: [519e4f1](https://github.com/redpanda-data/helm-charts/pull/117/commits/519e4f193b60c4e2410f2b1042c3eb2d958f5ea7)

File: redpanda/templates/post-upgrade.yaml (hunk 9 above)
Comment (redp01) [reply redpanda/templates/post-upgrade.yaml:42]: I also reuse the bootstrap.yaml file (through the mapped config) to push changes to the API

File: redpanda/templates/configmap.yaml (hunk 4 above)
Comment (redp01) [reply redpanda/templates/configmap.yaml]: Its pulling the old config format through lines 64-66

Comment (redp01) [reply redpanda/templates/configmap.yaml]: What this means is that if you have an older version of the values.yaml format, it is implied that you have transformed it to the cluster/tunables config format before you upgrade. If its lower than 22.1.1 then the old format should work.

File: redpanda/templates/configmap.yaml (hunk 5 above)
Comment (redp01) [reply redpanda/templates/configmap.yaml]: Fixed in [519e4f1](https://github.com/redpanda-data/helm-charts/pull/117/commits/519e4f193b60c4e2410f2b1042c3eb2d958f5ea7)

File: redpanda/templates/configmap.yaml (hunk 4 above)
Comment (redp01) [reply redpanda/templates/configmap.yaml]: We need to think about documenting the upgrade paths .......

File: redpanda/templates/post-upgrade.yaml (hunk 8 above)
Comment (redp01) [reply redpanda/templates/post-upgrade.yaml:46]: @vuldin - looks like that should be "sasl-enabled" only. Will update -  though logically would it be realistic for users to use this without tls?

File: redpanda/templates/configmap.yaml (hunk 4 above)
Comment (vuldin) [reply redpanda/templates/configmap.yaml]: We discussed this internally giving two options: 1) one `values.yaml` that works across all Redpanda versions and 2) requiring the user to specify an older Redpanda version as well as properly format the `config` section for that version.

We decided to go with option 2 since this will be less code in the helm chart (and potential for issues down the road), and it would be a more deliberate path for the user to choose an older version (considering we would rather support new Redpanda versions). We will provide documentation on using older Redpanda versions with the latest helm chart, which should help users with this.

File: redpanda/templates/configmap.yaml (hunk 6 above)
Comment (vuldin) [reply redpanda/templates/configmap.yaml:64]: After testing I see that these sections appear as `key: null` in the configmap output (rather than not existing). I think this is also related to a bug we discussed internally with using an older Redpanda version, where the cluster won't start with this branch.

File: redpanda/templates/post-upgrade.yaml (hunk 8 above)
Comment (vuldin) [reply redpanda/templates/post-upgrade.yaml:46]: We discussed internally and it seems we decided to continue allowing SASL without TSL.

File: redpanda/Chart.yaml [hunk 11]
```diff
@@ -21,7 +21,7 @@ maintainers:
     url: https://github.com/orgs/redpanda-data/people
 type: application
 version: 2.0.0-alpha.0
-appVersion: 21.11.16
+appVersion: 22.1.4
```
Comment (vuldin) [inline redpanda/Chart.yaml:24]:
```suggestion
appVersion: 22.1.5
```
This is a moving target, but as of now `22.1.5` is the latest.

File: redpanda/values.yaml [hunk 12]
```diff
@@ -34,7 +34,7 @@ commonLabels: {}
 image:
   repository: vectorized/redpanda
   # Redpanda version. This determines the installed version (not Chart.appVersion)
-  tag: v21.11.16
+  tag: v22.1.4
```
Comment (vuldin) [inline redpanda/values.yaml:37]:
```suggestion
  tag: v22.1.5
```

File: redpanda/values.yaml [hunk 13]
```diff
@@ -228,7 +228,7 @@ statefulset:
   replicas: 3
   updateStrategy:
     type: RollingUpdate
-  podManagementPolicy: Parallel
+  podManagementPolicy: OrderedReady
```
Comment (vuldin) [inline redpanda/values.yaml:231]: Did you notice any problems that led you to make this change? Do you see any differences in how brokers are aligned with nodes? Just interested in why this change now, and I do think it logically makes sense.

File: redpanda/Chart.yaml (hunk 11 above)
Comment (redp01) [reply redpanda/Chart.yaml]: resolved [46f6b62](https://github.com/redpanda-data/helm-charts/pull/117/commits/46f6b628b22162234c01501ab2e222b359aebc5e)

File: redpanda/values.yaml (hunk 12 above)
Comment (redp01) [reply redpanda/values.yaml]: resolved [46f6b62](https://github.com/redpanda-data/helm-charts/pull/117/commits/46f6b628b22162234c01501ab2e222b359aebc5e)

File: redpanda/templates/configmap.yaml [hunk 14]
```diff
@@ -89,6 +109,9 @@ data:
       - host:
           address: "{{ template "redpanda.fullname" $ }}-{{ . }}.{{ template "redpanda.internal.domain" $ }}"
           port: {{ $values.listeners.rpc.port }}
+{{- end }}
+{{- if not (eq (toYaml .Values.config.node) "null") }}
+  {{ toYaml .Values.config.node | indent 4 }}
```
Comment (vuldin) [inline redpanda/templates/configmap.yaml:114]: While testing with `21.11.17` I run into an issue due to this line, where it expects that `.Values.config.node` exists. The issue is that this section shouldn't exist at all in `21.x` configs.

A possible change to solve this would be the following:
```suggestion
{{- if hasKey .Values "config" }}
  {{- if not (eq (toYaml .Values.config.node) "null") }}
    {{ toYaml .Values.config.node | indent 4 }}
  {{- end }}
```

**Review by vuldin — CHANGES_REQUESTED** [review]
One issue while testing with `21.x`

File: redpanda/templates/configmap.yaml (hunk 14 above)
Comment (redp01) [reply redpanda/templates/configmap.yaml]: @vuldin - I have updated the configmap.yaml to be more tolerant to old versions with newer values.yaml files. New versions will require an explicit config change. Even if the ingest of values into central config is successful the declarative nature of the values file will not necessarily match the config in the central config. Note: upgrading values using the new format with an old values.yaml file was successful.

**Review by vuldin — APPROVED** [review]

File: redpanda/templates/tests/test-api-status.yaml [hunk 15]
```diff
@@ -15,7 +15,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */}}
 {{- /* TODO test fails if SASL is enabled */}}
```
Comment (vuldin) [inline redpanda/templates/tests/test-api-status.yaml:17]: Should this line be deleted after your changes to enable running SASL without TLS (non-blocker)?

File: redpanda/templates/configmap.yaml [hunk 16]
```diff
@@ -39,30 +54,49 @@ data:
...
     {{- end }}
     redpanda:
-{{- if .Values.auth.sasl.enabled }}
-  {{- with .Values.auth.sasl.users }}
+{{- if not (include "redpanda.semver" . | semverCompare ">=22.1.1") }}
+  {{- if .Values.auth.sasl.enabled }}
+    {{- with .Values.auth.sasl.users }}
       superusers:
-    {{ toYaml . | indent 8 }}
+    {{ toYaml . | indent 4 }}
```
Comment (vuldin) [inline redpanda/templates/configmap.yaml:61]: I believe L60-L61 should be the same as L38:
```suggestion
      superusers: ['{{ (first .).name }}']
```

**Review by vuldin — CHANGES_REQUESTED** [review]
I had one issue with getting a successful deployment when using Redpanda `v21.11.18`, related to properly reading the superuser from `values.yaml`.

**Comment by vuldin** [comment]
Resolves #109 . We will still want to create more tests around SASL outside of this PR.

File: redpanda/templates/configmap.yaml (hunk 16 above)
Comment (redp01) [reply redpanda/templates/configmap.yaml]: Fixed in [60a4927](https://github.com/redpanda-data/helm-charts/pull/117/commits/60a4927e61db5d117d538345f8b246d1fca73a0b) - I think this was a throw back to the alternative list notation.

File: redpanda/templates/NOTES.txt [hunk 17]
```diff
@@ -65,9 +67,17 @@ Delete the topic:
 
 {{- if and $anyTLS $anySASL }}
```
[the remark this answers is not in the document]
Comment (redp01) [reply redpanda/templates/NOTES.txt:88]: @vuldin - thanks. Fixed in [60a4927](https://github.com/redpanda-data/helm-charts/pull/117/commits/60a4927e61db5d117d538345f8b246d1fca73a0b)

File: redpanda/templates/NOTES.txt [hunk 18]
```diff
@@ -65,9 +67,17 @@ Delete the topic:
...
+Request the certificates:
+
+CERT=$(kubectl exec -n {{ .Release.Namespace }} {{ template "redpanda.name" . }}-0 -c redpanda -- cat /etc/tls/certs/kafka/ca.crt)
 
-CERT=$(kubectl exec -n {{ .Release.Namespace }} {{ template "redpanda.name" . }}-0 -c redpanda -- cat /etc/tls/certs/ca.crt)
+ADMINCERT=$(kubectl exec -n {{ .Release.Namespace }} {{ template "redpanda.name" . }}-0 -c redpanda -- cat /etc/tls/certs/admin/ca.crt)
+
+Create admin user:
+
+ kubectl -n redpanda run rpk -ti --rm --restart=Never --image {{ .Values.image.repository }}:{{ .Values.image.tag }} --command -- /bin/bash -c "rpk acl user create admin --api-urls {{ template "redpanda.name" . }}-0.{{ include "redpanda.internal.domain" . }}:{{ .Values.listeners.admin.port }} --user admin --password test --sasl-mechanism  SCRAM-SHA-256 --admin-api-tls-enabled --admin-api-tls-truststore <(echo '$ADMINCERT')"
```
[the remark this answers is not in the document]
Comment (redp01) [reply redpanda/templates/NOTES.txt]: @vuldin - fixed in [60a4927](https://github.com/redpanda-data/helm-charts/pull/117/commits/60a4927e61db5d117d538345f8b246d1fca73a0b)

**Review by vuldin — APPROVED** [review]
